2017/proposals/bansal-utkarsh.md

```diff
@@ -0,0 +1,628 @@
+Port testsuite to pytest
+========================
+
+
+Abstract
+========
+Testing is crucial to the development of any software and MDAnalysis currently uses nose to test their code. 
+Unfortunately, [nose](http://nose.readthedocs.io/en/latest/) is no longer under active development so the community has decided to shift over to [pytest](http://doc.pytest.org/en/latest/).
+The objective of this project is to implement this shift in a way that existing development work is not affected
+and to standardize and improve the existing test cases.
+
+
+Current State
+=============
+Currently the testsuite consists of 8 modules that have a total of 4731 test cases. Running this testsuite with 
+pytest (without any modifications) results in the following output:
+
+`1789 failed, 2630 passed, 52 skipped, 1610 pytest-warnings, 26 error`
+
+Also, it is to be noted that pytest misses out some (231 to be exact) test cases too. This is due to differences 
+between test discovery rules of nose and pytest.
+
+
+Technical Details
+=================
+
+Incompatible nose idioms
+------------------------
+
+1. Test classes with `__init__` methods are not collected by pytest.
+
+    **Example**
+
+    ```python
+    class TestGROReader(BaseReaderTest):
+      def __init__(self, reference=None):
+        if reference is None:
+        reference = GROReference()
+        super(TestGROReader, self).__init__(reference)
+
+      def test_flag_convert_lengths(self):
+        assert_equal(mda.core.flags['convert_lengths'], True,
+        "MDAnalysis.core.flags['convert_lengths'] should be True "
+        "by default")
+
+      def test_time(self):
+        u = mda.Universe(self.ref.topology, self.ref.trajectory)
+        assert_equal(u.trajectory.time, 0.0,
+        "wrong time of the frame")
+
+      def test_full_slice(self):
+        u = mda.Universe(self.ref.topology, self.ref.trajectory)
+        trj_iter = u.trajectory[:]
+        frames = [ts.frame for ts in trj_iter]
+        assert_equal(frames, np.arange(u.trajectory.n_frames))
+    ```
+
+
+    This class is not collected for tests because `__init__` is present
```

Comment:
I as a first step would try to get as much of the testsuite running under both pytest and nose.  If that means introducing some idoms that I want to remove later so be it. The rational is that this splits up the process into smaller units others can check easier (remember we do this in our free time while you will have your full day to work on this, so everything that makes our reviewing job easier is welcome).  You can later remove the TestCase inheritance if you wish.

You can check back with @jbarnoud and @richardjgowers wht they think about this.